lib/Doctrine/Common/Collections/Collection.php

```diff
@@ -212,6 +212,19 @@ public function next();
      */
     public function exists(Closure $p) : bool;
 
+    /**
+     * Returns the first element of this collection that satisfies the predicate p.
+     *
+     * @param Closure $p The predicate.
+     *
+     * @return mixed The first element respecting the predicate,
+     *               null if no element respects the predicate.
+     *
+     * @psalm-param Closure(TKey=, T=):bool $p
+     * @psalm-return T|null
```

Comment:
`findOne` returning null when result is not found makes it:
- Inconsistent with what `last()` and `first()` methods return. I think we should first solve this inconsistency before merging this PR
- Impossible to differentiate between missing result and collection containing `null` value

Reply:
I previously return `false` but got a review from @greg0ire https://github.com/doctrine/collections/pull/252#discussion_r475088024

I can return `false` for consistency, if you prefer. You just need to decide.

By the way, `last` and `first` returns false if not found ; but `get` return null if not found.

Reply:
Yeah I know we need to decide first. I'm fine with returning null, but we should first make existing code consistent so it returns null, otherwise we create more mess by merging this PR. I do have enough experience in software engineering to know better than to believe in "we will clean it up... later"

Reply:
I can see some logic in the current situation.

When we want to access to a value (like a `get`), `null` is return when not found.
When we want to perform an action 'like a `set`), `false` is return when a failure occurs.

`first` sets the index to the first element. If there is no element, it **cannot set**. Failure => `false`.
Same idea for `last`, `current`, `next`.

`key` gets the key of the element at the current iterator position. If there is no element, it **cannot get**. Nothing =>`null`.
Same idea for `get`. It would be the same idea for `findOne`.

The only remaining confusing method would be `remove` then.
It removes the element at a specific `key`, and return `null` if the element was already removed. By following the same logic, it should return `false` instead. (By the way, `removeElement` return `false` for the similar situation).
But [that's not the only issue I have with the `remove` method](https://github.com/doctrine/collections/issues/87#issuecomment-673672873): For consistency, I would prefer `remove` and `removeKey` instead of `removeElement` and `remove`.